parsers/IN_WE.py

```diff
@@ -0,0 +1,318 @@
+    "Sipat I": "coal",
+    "Sipat II": "coal",
+    "RGPPL": "gas",
+    "NSPCL": "coal",
+    "Mauda I": "coal",
+    "Mauda II": "coal",
+    "Sasan": "coal",
+    "CGPL": "coal",
+    "Solapur": "coal",
+    "Gadarwara": "coal",
+    "Lara": "coal",
+    "Khargone": "coal",
+}
+
+EXCHANGES_MAPPING = {
+    "WR-SR": "IN-SO->IN-WE",
+    "WR-ER": "IN-EA->IN-WE",
+    "WR-NR": "IN-NO->IN-WE",
+}
+
+KIND_MAPPING = {
+    "production": {"url": PRODUCTION_URL, "datetime_column": "lastUpdate"},
+    "exchange": {"url": EXCHANGE_URL, "datetime_column": "lastUpdate"},
+    "consumption": {"url": CONSUMPTION_URL, "datetime_column": "current_datetime"},
+}
+
+
+def fetch_data(
+    zone_key: str = "IN-WE",
+    kind: str = None,
+    session: Optional[Session] = None,
+    target_datetime: Optional[datetime] = None,
+    logger: Logger = getLogger(__name__),
+) -> dict:
+    """- get production data from wrldc.in
+    - filter all rows with same hour as target_datetime"""
+    r = session or Session()
+
+    dt_12_hour = arrow.get(target_datetime.strftime("%Y-%m-%d %I:%M")).datetime
+    payload = {"date": target_datetime.strftime("%Y-%m-%d")}
+
+    resp = r.post(url=KIND_MAPPING[kind]["url"], json=payload)
+
+    datetime_col = KIND_MAPPING[kind]["datetime_column"]
+    if resp.json():
+        data = json.loads(resp.json()["d"])
+        for item in data:
+            item[datetime_col] = datetime.strptime(
+                item[datetime_col], "%Y-%d-%m %H:%M:%S"
+            )
+            dt = arrow.get(item[datetime_col])
+            if dt.second >= 30:
+                item[datetime_col] = dt.shift(minutes=1).floor("minute").datetime
+            else:
+                item[datetime_col] = dt.floor("minute").datetime
+        df_data = pd.DataFrame(
+            [item for item in data if item[datetime_col].hour == dt_12_hour.hour]
+        )
+        return df_data
+    else:
+        raise ParserException(
+            parser="IN_WE.py",
+            message=f"{target_datetime}: {kind} data is not available",
+        )
+
+
+def format_production_data(
+    data: pd.DataFrame, zone_key: str, target_datetime: Optional[datetime]
+) -> dict:
+    """format production data:
+    - filters out correct datetimes (source data is 12 hour format)
+    - average all data points in the target_datetime hour
+    - map power plants
+    - sum production per mode"""
+    df_production = pd.DataFrame()
+    df_unique_dt = (
+        pd.DataFrame()
+    )  # for older datetimes datetimes are not the same in the AM and the PM which makes sorting harder
+    if len(data) == len(POWER_PLANT_MAPPING):
+        df_production = data.copy()
+    else:
+        for plant in set(data.State_Name):
+            df_plant = data.loc[data.State_Name == plant].copy()
+            for dt in set(df_plant.lastUpdate):
+                df_dt = df_plant.loc[df_plant.lastUpdate == dt]
+                if len(df_dt) > 1:
+                    if target_datetime.hour >= 12:
```

Comment:
Since target datetime can be None we should make sure target_datetime is not None before accesing it's properties.